Moved constructor logic into check to avoid unnecessary database connection

cgrabenstein opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | yes
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

Fetching the migrated versions in the constructor caused an unnecessary database connection as reported here: https://github.com/liip/LiipMonitorBundle/issues/259. This PR moves the logic from the constructor into the check method.

Commit:
Moved constructor logic into check to avoid unnecessary database connection

Signed-off-by: Christoph Grabenstein <christoph.grabenstein@qossmic.com>

cgrabenstein commented on src/Check/DoctrineMigration.php:
```diff
@@ -17,70 +17,96 @@
+            return $this->input->getAvailableVersions();
+        }
+
+        throw new LogicException(<<<'MESSAGE'
```
Not sure if this is a good way to handle this. Please suggest any improvements for handling the theoretical case that, despite the constructor validation, `$input` is neither `DependencyFactory` not `Configuration`

Ocramius replied:
If the constructor parameter type is properly specified, no check is really needed

Ocramius replied:
Since `$this->input` is either a `DependencyFactory` or a `Configuration`, you can remove the second `if ()` completely, and just do `return $this->input->getAvailableVersions();` :+1:

cgrabenstein commented:
@weierophinney @Ocramius just a ping to check if there's anything I can help you with so this can be reviewed and merged more quickly.

Ocramius reviewed (changes requested):
This change really needs a unit test verifying that, if only the constructor was used, no connection attempt was performed.

Commit:
Refactoring

Signed-off-by: Christoph Grabenstein <christoph.grabenstein@qossmic.com>

Commit:
Added unit tests

Signed-off-by: Christoph Grabenstein <christoph.grabenstein@qossmic.com>

cgrabenstein commented:
@Ocramius Added a test case for version 2 and 3 each.

Ocramius commented on src/Check/DoctrineMigration.php:
```diff
@@ -17,70 +17,94 @@
+            return $this->getMigratedVersionsFromDependencyFactory($this->input);
+        }
+
+        if ($this->input instanceof Configuration) {
```
This conditional can be removed, see https://github.com/laminas/laminas-diagnostics/pull/20/files#r569752752

Ocramius commented:
@cgrabenstein thanks for adapting this! The last bit missing is really just some refactoring, but then we should be through.

Also, can you attempt a rebase, or should I do it for you? That will highlight some `vimeo/psalm` violations, I think.

Commit:
Removed unnecessary ifs

Signed-off-by: Christoph Grabenstein <christoph.grabenstein@qossmic.com>

Commit:
Added return types

Signed-off-by: Christoph Grabenstein <christoph.grabenstein@qossmic.com>

cgrabenstein commented:
@Ocramius I did the rebase and refactoring. The following psalm issues showed up and I'm not sure how to handle those, as they are related to the fact that the class is handling two doctrine versions simultaneously. 

```
ERROR: MixedInferredReturnType - src/Check/DoctrineMigration.php:65:16 - Could not verify return type 'array<array-key, Doctrine\Migrations\Version\Version>' for Laminas\Diagnostics\Check\DoctrineMigration::getAvailableVersions (see https://psalm.dev/047)
     * @return Version[]


ERROR: MixedReturnStatement - src/Check/DoctrineMigration.php:73:16 - Could not infer a return type (see https://psalm.dev/138)
        return $this->input->getAvailableVersions();


ERROR: UndefinedMethod - src/Check/DoctrineMigration.php:73:30 - Method Doctrine\Migrations\Configuration\Configuration::getAvailableVersions does not exist (see https://psalm.dev/022)
        return $this->input->getAvailableVersions();


ERROR: MixedInferredReturnType - src/Check/DoctrineMigration.php:77:16 - Could not verify return type 'array<array-key, Doctrine\Migrations\Version\Version>' for Laminas\Diagnostics\Check\DoctrineMigration::getMigratedVersions (see https://psalm.dev/047)
     * @return Version[]


ERROR: MixedReturnStatement - src/Check/DoctrineMigration.php:85:16 - Could not infer a return type (see https://psalm.dev/138)
        return $this->input->getMigratedVersions();


ERROR: UndefinedMethod - src/Check/DoctrineMigration.php:85:30 - Method Doctrine\Migrations\Configuration\Configuration::getMigratedVersions does not exist (see https://psalm.dev/022)
        return $this->input->getMigratedVersions();
```
Shall I just add them to the baseline?

Ocramius commented:
@cgrabenstein no worries - it's kinda expected to have `UndefinedMethod` calls here, since the dependency is not actually installed, and we're duck-typing our way through.

Ocramius committed:
Baselining errors caused by missing `doctrine/migrations` dependency (not installed)

`doctrine/migrations` is supported in multiple versions/flavors, and therefore we duck-type our way
through a setup here.

This is sub-optimal, and needs a long-term solution later on, but for now, we roll with this approach,
baselining errors related to unknown classes.

A better approach would really be to split up the package into multiple "checks", each to be installed
individually, but that leads to terrible usability/maintainability too.

Discussion around this is deferred.

Ocramius reviewed (approved):
Rebased, baselined errors, and merging. Thanks @cgrabenstein!
